src/lib/optimizer/optimizer.cpp

```diff
@@ -149,118 +285,30 @@ std::shared_ptr<AbstractLQPNode> Optimizer::optimize(
   }
 
   // Remove LogicalPlanRootNode
-  auto optimized_node = root_node->left_input();
+  const auto optimized_node = root_node->left_input();
   root_node->set_left_input(nullptr);
 
   return optimized_node;
 }
 
 void Optimizer::validate_lqp(const std::shared_ptr<AbstractLQPNode>& root_node) {
-  // If you can think of a way in which an LQP can be corrupt, please add it!
-
-  // First, collect all LQPs (the main LQP and all subqueries)
+  // First, collect all LQPs (the main LQP and all uncorrelated subqueries).
   auto lqps = std::vector<std::shared_ptr<AbstractLQPNode>>{root_node};
-  auto subquery_expressions_by_lqp = collect_lqp_subquery_expressions_by_lqp(root_node);
-  for (const auto& [lqp, subquery_expressions] : subquery_expressions_by_lqp) {
+  auto subquery_expressions_by_lqp = collect_lqp_subquery_expressions_by_lqp(root_node, true);
+  for (const auto& [lqp, _] : subquery_expressions_by_lqp) {
     lqps.emplace_back(lqp);
   }
 
-  std::unordered_map<std::shared_ptr<const AbstractLQPNode>, std::unordered_set<std::shared_ptr<const AbstractLQPNode>>>
-      nodes_by_lqp;
+  auto nodes_by_lqp = LQPNodesByLQP{};
+  // Second, assign each LQPNode to its LQP. Uncorrelated subqueries are treated like normal LQPNodes since their output
+  // can also be used multiple times.
```

Comment:
Correlated subquery LQPs must not be reused, they are completely managed by the ExpressionEvaluator. Thus, they cannot have multiple "outputs", i.e., must not be part of multiple LQPSubqueryExpressions. For a long time, we were a bit shy about subqueries and treated uncorrelated subqueries the same, which was unnecessary. Since the results of uncorrelated subquery LQPs do not depend on the value of a correlated parameter, we can use them in a thousand subquery expressions if we like, which is exactly what is the case for all "regular" LQP nodes, which can have an unlimited number of outputs. Recently, we started to treat uncorrelated subqueries more like regular LQPNodes/Operators, we schedule them like any other task, etc. So no need to treat them differently here.

tl;dr: `Since the results of uncorrelated subquery LQPs do not depend on correlated parameters, we can use them in multiple LQPSubqueryExpressions and treat them like normal LQPNodes. However, correlated subqueries must not share nodes with any other (sub-)LQP.`?

Reply:
I like the comment.

Should we also name that the expr eval owns the correlated LQPs and deep-copies them?